### PR TITLE
WIP: Titanic example

### PR DESCRIPTION
**Pull request by yangarbiter**
@ianlini

### CODE REVIEW

File: examples/titanic/.feagenrc/config.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+generator_class: titanic.TitanicFeatureGenerator
+data_bundles_dir: data_bundles
+
+# The additional arguments that will be given when initiating the data generator
+# object.
+generator_kwargs:
+  h5py_hdf_path:
+    global_data.h5
```
Comment (ianlini) [inline examples/titanic/.feagenrc/config.yml]: I've changed this

File: examples/titanic/model.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import pandas as pd
+from sklearn.ensemble import RandomForestClassifier
+
+def load_feature_run_model(bundle_hdf_path, prediction_csv_path):
+    bundle_f = h5py.File(bundle_hdf_path, 'r')
+
+    is_valid = np.array(bundle_f['info']['is_valid'])
+    is_test = np.array(bundle_f['info']['is_test'])
+    passenger_id = np.array(bundle_f['id']['passenger_id'])
+    label = np.array(bundle_f['label']['label'])
```
Comment (ianlini) [inline examples/titanic/model.py]: use `.value`

File: examples/titanic/model.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    # concated
+    feature = np.array(bundle_f['features'])
```
Comment (ianlini) [inline examples/titanic/model.py]: `.value`

File: examples/titanic/model.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    bundle_f.close()
```
Comment (ianlini) [inline examples/titanic/model.py]: use `with`

File: examples/titanic/model.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    train_filter = (np.bitwise_and(is_valid == 0, is_test == 0))
+    valid_filter = (np.bitwise_and(is_valid == 1, is_test == 0))
+    test_filter = (is_test == 1)
```
Comment (ianlini) [inline examples/titanic/model.py]: bad code

File: examples/titanic/model.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    df = pd.DataFrame(prediction, columns=['Survived'],
+                      index=passenger_id[test_filter])
+    df.index.rename('PassengerId')
+    df.to_csv(prediction_csv_path)
+
+
+if __name__ == '__main__':
+    load_feature_run_model(
+        os.path.join(os.path.dirname(__file__), 'data_bundles', 'feature01.h5'),
+        "prediction.csv")
```
Comment (ianlini) [inline examples/titanic/model.py]: write variable name

File: examples/titanic/model.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    valid_filter = (np.bitwise_and(is_valid == 1, is_test == 0))
+    test_filter = (is_test == 1)
+
+    ##############
+    # validation #
+    ##############
+    clf = RandomForestClassifier()
+    clf.fit(feature[train_filter], label[train_filter])
+    print('validation score:',
+          clf.score(feature[valid_filter], label[valid_filter]))
```
Comment (ianlini) [inline examples/titanic/model.py:29]: You should say what score is

File: examples/titanic/titanic.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    @will_generate('h5py', 'passenger_id')
+    def gen_passenger_id(self, data):
+        data_df = data['data_df']
+        return {'passenger_id': data_df.index.values}
+
+    @require('data_df')
+    @will_generate('h5py', 'is_test')
+    def gen_is_test(self, data):
+        temp = data['data_df']['Survived'].values
+        return {'is_test': (temp == -1)}
```
Comment (ianlini) [inline examples/titanic/titanic.py]: no `temp`

File: examples/titanic/titanic.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    @will_generate('h5py', 'is_valid')
+    def gen_is_validation(self, data):
+        from sklearn.model_selection import train_test_split
+        import numpy as np
+        data_df = data['data_df']
+        df = pd.DataFrame(0, index=data_df.index, columns=['is_valid'], dtype=bool)
+        random_state = np.random.RandomState(1126)
+        _, valid_id = train_test_split(
+            data_df.index, test_size=0.3, random_state=random_state)
+        df.loc[valid_id, 'is_valid'] = 1
```
Comment (ianlini) [inline examples/titanic/titanic.py]: use bool

Comment (ianlini) [reply examples/titanic/titanic.py]: or `data_df.index.isin(valid_id)`

File: examples/titanic/titanic.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        data_df = data['data_df']
+        return {'label': data_df['Survived'].values}
+
+    @require('data_df')
+    @will_generate('h5py', 'pclass')
+    def gen_pclass(self, data):
+        from sklearn.preprocessing import OneHotEncoder
+        data_df = data['data_df']
+        pclass = np.array(data_df['Pclass'].values)
+        pclass[np.isnan(pclass)] = 4 # unknown as a class
```
Comment (ianlini) [inline examples/titanic/titanic.py]: use pandas to do these operations

File: examples/titanic/titanic.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        return {'pclass': OneHotEncoder(sparse=False)
+                          .fit_transform(pclass.reshape((-1, 1)))}
```
Comment (ianlini) [inline examples/titanic/titanic.py]: think more 
reshape can be eliminated if you add two character

File: examples/titanic/titanic.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        data_df = data['data_df']
+        return {'family_size': (data_df['SibSp'] + data_df['Parch']).values}
+
+    @require('data_df')
+    @will_generate('h5py', ['age', 'sibsp'])
+    def gen_age_sibsp(self, data):
+        data_df = data['data_df']
+        # clean up age data
+        age = data_df['Age'].values
+        age[np.isnan(age)] = np.mean(age[~np.isnan(age)])
```
Comment (ianlini) [inline examples/titanic/titanic.py]: use pandas

File: examples/titanic/titanic.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+                'sibsp': data_df['SibSp'].values}
+
+def generate_titanic_features(h5py_hdf_path, bundle_hdf_path):
+    label_list = ['label']
+    info_list = ['is_valid', 'is_test']
+    id_list = ['passenger_id']
+    feature_list = ['family_size', 'sibsp', 'age', 'pclass']
+    generator = TitanicFeatureGenerator(h5py_hdf_path,
+        os.path.join(os.path.abspath(__file__), 'data', 'train.csv'),
+        os.path.join(os.path.abspath(__file__), 'data', 'test.csv'))
```
Comment (ianlini) [inline examples/titanic/titanic.py]: why not move these to main?

File: examples/titanic/titanic.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+                        'info': info_list,
+                        'id': id_list,
+                        'features': feature_list}
+    structure_config = {'features': {'concat': True}}
+    generator.bundle(bundle_structure, data_bundle_hdf_path=bundle_hdf_path,
+            structure_config=structure_config)
+
+def main():
+    generate_titanic_features(
+        os.path.join(os.path.dirname(__file__), 'data', 'global_feature.h5'),
```
Comment (ianlini) [inline examples/titanic/titanic.py]: the naming convention of `global_feature.h5` has been changed

**Review by ianlini — CHANGES_REQUESTED** [review]

File: examples/titanic/titanic.py [hunk 15]
```diff
@@ -59,11 +59,11 @@ def gen_label(self, data):
     @will_generate('h5py', 'pclass')
     def gen_pclass(self, data):
         from sklearn.preprocessing import OneHotEncoder
-        data_df = data['data_df'].copy() # it is mutable
+        data_pclass = data['data_df'][['Pclass']]
         # set unknown as a class
-        data_df.loc[data_df['Pclass'].isnull(), 'Pclass'] = 4
-        return {'pclass': OneHotEncoder(sparse=False).fit_transform(
-                            data_df['Pclass'].values.reshape((-1, 1)))}
+        data_pclass = data_pclass.fillna(4)
```
Comment (ianlini) [inline examples/titanic/titanic.py]: use `inplace=True`

File: examples/titanic/titanic.py [hunk 16]
```diff
@@ -74,7 +74,7 @@ def gen_family_size(self, data):
     @require('data_df')
     @will_generate('h5py', ['age', 'sibsp'])
     def gen_age_sibsp(self, data):
-        data_df = data['data_df'].copy()
+        data_df = data['data_df'].copy() # mutable
```
Comment (ianlini) [inline examples/titanic/titanic.py]: what mutable?

**Review by ianlini — CHANGES_REQUESTED** [review]

File: examples/titanic/titanic.py (hunk 16 above)
Comment (yangarbiter) [reply examples/titanic/titanic.py]: data['data_df'] is mutable so we should copy before modifying it?

Comment (ianlini) [reply examples/titanic/titanic.py]: why did you think people will know what you are thinking by just saying "mutable"?

File: examples/titanic/titanic.py [hunk 17]
```diff
@@ -61,7 +61,7 @@ def gen_pclass(self, data):
         from sklearn.preprocessing import OneHotEncoder
         data_pclass = data['data_df'][['Pclass']]
         # set unknown as a class
-        data_pclass = data_pclass.fillna(4)
+        data_pclass = data_pclass.fillna(4, inplace=True)
```
Comment (ianlini) [inline examples/titanic/titanic.py]: `data_pclass.fillna(4, inplace=True)`

**Review by ianlini — CHANGES_REQUESTED** [review]

**Review by ianlini — APPROVED** [review]

**Pull request merged**